src/Microsoft.AspNetCore.SignalR.Core/Internal/AsyncEnumeratorAdapters.cs

```diff
@@ -39,15 +43,14 @@ public static IAsyncEnumerator<object> FromObservable<T>(IObservable<T> observab
             // Dispose the subscription when the token is cancelled
             cancellationToken.Register(state => ((IDisposable)state).Dispose(), subscription);
 
-            return channel.In.GetAsyncEnumerator(cancellationToken);
+            return GetAsyncEnumerator(channel.In, cancellationToken);
         }
 
         public static IAsyncEnumerator<object> FromChannel(object readableChannelOfT, Type payloadType, CancellationToken cancellationToken)
         {
-            var enumerator = readableChannelOfT
-                .GetType()
-                .GetRuntimeMethod("GetAsyncEnumerator", new[] { typeof(CancellationToken) })
-                .Invoke(readableChannelOfT, new object[] { cancellationToken });
+            var enumerator = _getAsyncEnumeratorMethod
```

Comment:
I wonder if we really need to use reflection here...

Reply:
We do because we only know the T of the channel via `payloadType` hence the `MakeGenericMethod`

Reply:
I guess you are right - you might be able push it up a bit but eventually you will have to set the generic argument.